### PR TITLE
Homework 6. ATM management

### PR DESCRIPTION
**Pull request by DmitryAstreiko**
изображение

изображение

 1. Delegate and events. Collection.
 2. The task was assigned 24.12.2020, the task was completed between 28.12.2020 and 29.12.2020.
 3. None

### CODE REVIEW

File: src/Astreiko.Homework6/Atm.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            EuroToByn = 3.11M;
+        }
+
+        /// <summary>
+        /// Withdraw amount with balance
+        /// </summary>
+        /// <param name="delSum"></param>
+        public void GetCash(decimal delSum)
+        {
+            Predicate<decimal> checkSum = (x) => (x > _currentBalance);
```
Comment (diman84) [inline src/Astreiko.Homework6/Atm.cs]: This looks odd ) even for demonstration purposes, it adds excessive complexity

File: src/Astreiko.Homework6/Program.cs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                switch (inputChoose)
+                {
+                    case "a":
+                        currentAtm.PutCash(GetSum("Enter sum to add [BYN] : "));
+                        ShowMainMenu();
+                        break;
+                    case "w":
+                        ShowMenuCurrency(currentAtm.EuroToByn, currentAtm.UsdToByn);
+                        WorkWithSum(currentAtm);
+                        ShowMainMenu();
```
Comment (diman84) [inline src/Astreiko.Homework6/Program.cs]: this is duplicated across case blocks, can be refactored

File: src/Astreiko.Homework6/Program.cs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+namespace Astreiko.Homework6
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var currentAtm = new Atm();
+
+            currentAtm.AddBalanceHandler += CurrentAtm_AddBalanceHandler;
```
Comment (diman84) [inline src/Astreiko.Homework6/Program.cs]: Events usually named as a verb in passive or with On prefix to make it more describing and readable

**Review by diman84 — COMMENTED** [review]
Generally good

**Comment by DmitryAstreiko** [comment]
HI.
My question is, what do you think of declaring variables (see screen below)?
изображение
or
изображение

I think that in first variant variables will be create variables in stack as many times as we will be use section while. 

In second variant we will create variables one time and section while will be use this variables many times. 

May be the cost of allocating memory will be less in the second variant.

**Comment by diman84** [comment]
> HI.
> My question is, what do you think of declaring variables (see screen below)?
> изображение
> or
> изображение
> 
> I think that in first variant variables will be create variables in stack as many times as we will be use section while.
> 
> In second variant we will create variables one time and section while will be use this variables many times.
> 
> May be the cost of allocating memory will be less in the second variant.

Hi, thanks for interesting question. The compiler is too smart to allocate something more that required.

So, for the following 3 methods:

```
       private static decimal GetNumber1()
        {
            decimal number;
            bool result;
            while (true)
            {
                result = decimal.TryParse(Console.ReadLine(), out number);
                if (result)
                {
                    return number;
                }                
                Console.WriteLine("Enter correct number.");
            }
        }

        private static decimal GetNumber2()
        {
            while (true)
            {
                decimal number;
                bool result;
                result = decimal.TryParse(Console.ReadLine(), out number);
                if (result)
                {
                    return number;
                }                
                Console.WriteLine("Enter correct number.");
            }
        }

        private static decimal GetNumber3()
        {
            while(true)
            {   
                if (decimal.TryParse(Console.ReadLine(), out var number))
                    return number;
                Console.WriteLine("Enter correct number.");
            }
        }
```

the IL code that is generated is just the same:

```
   .method private hidebysig static valuetype [System.Runtime]System.Decimal
     GetNumber1() cil managed
   {
    .maxstack 2
    .locals init (
      [0] valuetype [System.Runtime]System.Decimal number
    )
    // start of loop, entry point: IL_0000    
      IL_0000: call         string [System.Console]System.Console::ReadLine()
      IL_0005: ldloca.s     number
      IL_0007: call         bool [System.Runtime]System.Decimal::TryParse(string, valuetype [System.Runtime]System.Decimal&)

      IL_000c: brfalse.s    IL_0010

      IL_000e: ldloc.0      // number
      IL_000f: ret

      IL_0010: ldstr        "Enter correct number."
      IL_0015: call         void [System.Console]System.Console::WriteLine(string)

      IL_001a: br.s         IL_0000
    // end of loop

  } // end of method Program::GetNumber1

  .method private hidebysig static valuetype [System.Runtime]System.Decimal
    GetNumber2() cil managed
  {
    .maxstack 2
    .locals init (
      [0] valuetype [System.Runtime]System.Decimal number
    )

    // start of loop, entry point: IL_0000
      IL_0000: call         string [System.Console]System.Console::ReadLine()
      IL_0005: ldloca.s     number
      IL_0007: call         bool [System.Runtime]System.Decimal::TryParse(string, valuetype [System.Runtime]System.Decimal&)

      IL_000c: brfalse.s    IL_0010

      IL_000e: ldloc.0      // number
      IL_000f: ret

      IL_0010: ldstr        "Enter correct number."
      IL_0015: call         void [System.Console]System.Console::WriteLine(string)

      IL_001a: br.s         IL_0000
    // end of loop

  } // end of method Program::GetNumber2

  .method private hidebysig static valuetype [System.Runtime]System.Decimal
    GetNumber3() cil managed
  {
    .maxstack 2
    .locals init (
      [0] valuetype [System.Runtime]System.Decimal number
    )

    // start of loop, entry point: IL_0000
      IL_0000: call         string [System.Console]System.Console::ReadLine()
      IL_0005: ldloca.s     number
      IL_0007: call         bool [System.Runtime]System.Decimal::TryParse(string, valuetype [System.Runtime]System.Decimal&)
      IL_000c: brfalse.s    IL_0010

      IL_000e: ldloc.0      // number
      IL_000f: ret

      IL_0010: ldstr        "Enter correct number."
      IL_0015: call         void [System.Console]System.Console::WriteLine(string)

      IL_001a: br.s         IL_0000
    // end of loop

  } // end of method Program::GetNumber3
```
All three methods has same stack size, local variables and consists of 9 identical commands.

**Comment by DmitryAstreiko** [comment]
OOOOO, third method is excellent. 

I am sorry, that I wasting your times, but I have one variant check cost of allocating memory.

Your examples are very good. 
You entered 1 times correct number in every method, but maybe results will be different if we enter bad number (first step), bad number (second step) and correct number (third step) in every method.

@diman84 if you will say me how you see IL code, I do it myself.

Thank you.

**Comment by diman84** [comment]
> OOOOO, third method is excellent.
> 
> I am sorry, that I wasting your times, but I have one variant check cost of allocating memory.
> 
> Your examples are very good.
> You entered 1 times correct number in every method, but maybe results will be different if we enter bad number (first step), bad number (second step) and correct number (third step) in every method.
> 
> @diman84 if you will say me how you see IL code, I do it myself.
> 
> Thank you.

Since the output of C# compiler is IL library and it is identical for all variants there won't be any difference and there won't be any additional allocations on stack. 
To view IL code of .net assemblies the most common tool is ILSpy, it is lightweight and free, I used dotPeek from JetBrains because I have reSharper setup.

**Review by diman84 — APPROVED** [review]

**Pull request merged**